DSharpPlus/Entities/Channel/DiscordChannel.cs

```diff
@@ -624,34 +624,72 @@ public async Task<ThreadQueryResult> ListJoinedPrivateArchivedThreadsAsync(DateT
+    /// Deletes multiple messages if they are less than 14 days old.  If they are older, none of the messages will be deleted and you will receive a <see cref="ArgumentException"/> error.
     /// </summary>
     /// <param name="messages">A collection of messages to delete.</param>
     /// <param name="reason">Reason for audit logs.</param>
-    /// <returns></returns>
-    /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageMessages"/> permission.</exception>
-    /// <exception cref="NotFoundException">Thrown when the channel does not exist.</exception>
-    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
-    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
-    public async Task DeleteMessagesAsync(IEnumerable<DiscordMessage> messages, string reason = null)
+    /// <remarks>One api call per 100 requests</remarks>
+    public async Task<int> DeleteMessagesAsync(IReadOnlyList<DiscordMessage> messages, string reason = null)
     {
-        // don't enumerate more than once
-        ulong[] msgs = messages.Where(x => x.Channel.Id == this.Id).Select(x => x.Id).ToArray();
-        if (messages == null || !msgs.Any())
+        ArgumentNullException.ThrowIfNull(messages, nameof(messages));
+
+        List<ulong> messagesList = [];
+        foreach (DiscordMessage message in messages)
+        {
+            if (message.ChannelId != this.Id)
+            {
+                throw new ArgumentException(
+                    "You can only delete messages from the channel this channel object is representing.");
+            }
+            else if (message.Timestamp < DateTimeOffset.UtcNow.AddDays(-14))
+            {
+                throw new ArgumentException("You can only delete messages that are less than 14 days old.");
+            }
+
+            messagesList.Add(message.Id);
+        }
+
+        if (messagesList.Count == 0)
         {
             throw new ArgumentException("You need to specify at least one message to delete.");
         }
+        else if (messagesList.Count < 2)
+        {
+            await this.Discord.ApiClient.DeleteMessageAsync(this.Id, messagesList.Single(), reason);
+            return 1;
+        }
 
-        if (msgs.Count() < 2)
+        int count = 0;
+        for (int i = 0; i < messagesList.Count; i += 100)
         {
-            await this.Discord.ApiClient.DeleteMessageAsync(this.Id, msgs.Single(), reason);
-            return;
+            //first iter 0 - 99 -> 100 Messages
+            //second iter 100 - 199 -> 100 Messages
+            await this.Discord.ApiClient.DeleteMessagesAsync(this.Id,
+                messagesList[i .. Math.Min(messagesList.Count - 1, i + 99)], reason);
+            count += Math.Min(messagesList.Count - 1, i + 99) - i;
```

Comment:
Why this over `Math.Min(messagesList.Count, i + 100)`?

Reply:
+ 99 adds an overlap which will result in a exception. Read comment above

Reply:
Tested or suspected? In my testing there were never more than 100 messages selected.

Reply:
Math. First api call uses 0-100 and next would be 100-200. So 100 would be included twice which is a NotFound from discord afaik

Reply:
I'll take your word for it

Reply:
if you still have your tests, could you quickly test if each id is only used once?

Reply:
Sure

Reply:
Duplicates were found:
```cs
List<string> messageIds = [];
for (int i = 0; i < messagesList.Count; i += 100)
{
    IEnumerable<string> partMessageIds = messagesList[i..Math.Min(messagesList.Count, i + 100)];
    messageIds.AddRange(partMessageIds);
    await Task.Delay(0);
}

if (messageIds.Count != messageIds.Distinct().Count())
{
    throw new ArgumentException("You cannot delete the same message multiple times.");
}
```

Reply:
so my asumption was right?

Reply:
Yes